nav2_collision_monitor/src/polygon_velocity.cpp

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2022 Dexory
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/polygon_velocity.hpp"
+
+
+namespace nav2_collision_monitor
+{
+
+PolygonVelocity::PolygonVelocity(
+  const std::vector<Point> & poly,
+  const std::string & polygon_name,
+  const double & linear_max,
+  const double & linear_min,
+  const double & direction_max,
+  const double & direction_min,
+  const double & theta_max,
+  const double & theta_min)
+: poly_(poly), polygon_name_(polygon_name), linear_max_(linear_max), linear_min_(linear_min),
+  direction_max_(direction_max), direction_min_(direction_min),
+  theta_max_(theta_max), theta_min_(theta_min)
+{
+  RCLCPP_INFO(logger_, "[%s]: Creating PolygonVelocity", polygon_name_.c_str());
+}
+
+
+PolygonVelocity::~PolygonVelocity()
+{
+  RCLCPP_INFO(logger_, "[%s]: Destroying PolygonVelocity", polygon_name_.c_str());
+  poly_.clear();
+}
+
+bool PolygonVelocity::isInRange(const Velocity & cmd_vel_in)
+{
+  const double twist_linear = std::hypot(cmd_vel_in.x, cmd_vel_in.y);
+
+  // check if direction in angle range(min -> max)
+  double direction = std::atan2(cmd_vel_in.y, cmd_vel_in.x);
+  bool direction_in_range;
+  if (direction_min_ <= direction_max_) {
+    direction_in_range = (direction >= direction_min_ && direction <= direction_max_);
+  } else {
+    direction_in_range = (direction >= direction_min_ || direction <= direction_max_);
```

Comment:
Wouldn't this be `direction >= direction_max_ || direction <= direction_min_`?

This if/else for the directions I think may need some work, but I'm in currently in a position to do a table of all the options to validate. Mostly a note for myself / @kaichie to potentially look at this again. Especially when passing the PI line or when dealing with negatives (including potentially making sure that the direction settings are in the correct -PI, PI frame)

Reply:
Please see the [comment](https://github.com/ros-planning/navigation2/pull/3708#discussion_r1292861566) here. I think `direction >= direction_max_ || direction <= direction_min_` will not cover the current assumption. But we can discuss the concept first and see if there is a better way to encompass the direction range.

Reply:
As mentioned in the previous comment, the polygon's covered direction always spans from `direction_min` to `direction_max` in clockwise. 
The direction is calculated with `std::atan2(y, x)`, resulting in an output range of `-π` to `π`.
The direction should be defined as follow:
![reference](https://github.com/ros-planning/navigation2/assets/37255935/07071a13-50f1-4869-9e28-7aa13f725831)

To validate the logic, the following examples are considered:
![direction_ranges](https://github.com/ros-planning/navigation2/assets/37255935/4ab7300b-c01d-4669-859e-5de140ca45ce)
![table1](https://github.com/ros-planning/navigation2/assets/37255935/3e5b370a-ecf2-4f2b-88ae-67231256b033)

In cases such as 5 and 6, a user could potentially use -π instead of π; however the direction verification should remains `True`:
![table2](https://github.com/ros-planning/navigation2/assets/37255935/8420bcef-43f8-45b1-a8bf-7f58dc8463f2)

Here's a basic Python script I used to test the logic:
```python
#!/usr/bin/env python3

import sys

def check_direction_in_range(direction, direction_min, direction_max):
    if direction_min <= direction_max:
        direction_in_range = direction >= direction_min and direction <= direction_max
    else:
        direction_in_range = direction >= direction_min or direction <= direction_max
    return direction_in_range

if len(sys.argv) != 4:
    print("Invalid usage. Usage: python3 script.py direction direction_min direction_max")
    sys.exit()

direction, direction_min, direction_max = map(float, sys.argv[1:4])
result = check_direction_in_range(direction, direction_min, direction_max)
print("Direction is in range:", result)
```

Reply:
> So if you were to have something from just forward to left, you'd do it 0.5 PI as min and 0 as max? That doesn't make sense... That's counter-clockwise to do that in a logical way. I assume that's just a typo.

Apologies for the typo; it should be counter-clockwise. Therefore, the forward and left would be 0 as min and 0.5 PI as max.

> Then if you did it always counter-clockwise, then having 0.75 PI be the min and -0.75PI be the max equally makes little sense. I think you need to abandon this dirctionality logic of defining the range.

Okay I believe there might be some confusion regarding the parameter names. Technically, we are trying to determine if an angle is between two angle(range of -PI to PI). A more intuitive way to refer to these parameter names could be `direction_start_angle` and `direction_end_angle`. This naming would help avoid the confusion of having to set the `min` angle numerically greater than the `max` angle. However, we still need the underlying assumption of `direction_start_angle` to `direction_end_angle` is in the counter-clockwise direction. How does this sounds?

As for the logic, I've come across a couple of interesting reads([here](https://stackoverflow.com/a/29721295) and another one [here](https://stackoverflow.com/questions/11406189/determine-if-angle-lies-between-2-other-angles)) that have a similar implementation. I will try if it is possible to `sort them at configuration time` as mentioned. Totally open for any ideas and suggestions!

> But that's not a objective thought, but my subjective opinion. I just see lots of people complaining or misconfiguring how the positive value is somehow the "minimum" where as the negative value is the "maximum" by the range definitions.

I understand your perspective, lets avoid this :100:

Reply:
Perhaps this is OK as long as we document it explicitly. You have that nice graphic now so I suppose if that was linked in the readme/config guide to explain how to configure it, there's little excuse for people to mess it up. I know alot of people **don't** read the docs which is what I was trying to combat against, but I suppose if there's a visual and clear explanation I can point them to, then the overhead on me answering questions is limited to a link. 

Nothing's wrong with your method, I just try to find the most bullet proof way of handling it as possible to head-off user errors / issues. 

For the parameters, I think renaming also helps. Not required but you could make it an array param `translational_direction: [0.5, 0.6]` (or direction or linear direction, I think just being consistent with the other things is key) instead of specifying min/max or start/end explicitly. That might actually help by forcing folks into the docs to understand what it means :-)